docs/design/environments-implementation.md

```diff
@@ -0,0 +1,193 @@
+
+## Kernel Implementation
+
+After investigating different alternatives (such as pipenv, venv, etc.), we decided to use Conda environment which nicely replaces Python virtual env, pip, and can also support other languages. More details can be found at: https://medium.com/@krishnaregmi/pipenv-vs-virtualenv-vs-conda-environment-3dde3f6869ed
+
+When once Conda, users can easily add, remove dependency of a Conda environment. User can also easily export environment to yaml file.
+
+The yaml file of Conda environment by using `conda env export` looks like: 
+
+```
+name: base
+channels:
+  - defaults
+dependencies:
+  - _ipyw_jlab_nb_ext_conf=0.1.0=py37_0
+  - alabaster=0.7.12=py37_0
+  - anaconda=2020.02=py37_0
+  - anaconda-client=1.7.2=py37_0
+  - anaconda-navigator=1.9.12=py37_0
+  - anaconda-project=0.8.4=py_0
+  - applaunchservices=0.2.1=py_0
+```
+
+Including Conda kernel, the environment object may look like: 
+
+```
+name: "my_submarine_env",
+    vm-image: "...",
+    docker-image: "...", 
+    kernel: 
+      name: team_default_python_3.7
+      channels:
+        - defaults
+      dependencies:
+        - _ipyw_jlab_nb_ext_conf=0.1.0=py37_0
+        - alabaster=0.7.12=py37_0
+        - anaconda=2020.02=py37_0
+        - anaconda-client=1.7.2=py37_0
+        - anaconda-navigator=1.9.12=py37_0
+```
+
+When launch a new experiment / notebook session using the `my_submarine_env`, submarine server will use defined Docker image, and Conda kernel to launch of container. 
+
+## Storage of Environment 
+
+Environment of Submarine is just a simple text file, so it will be persisted in Submarine metastore, which is ideally a Database. 
+
+Docker image is stored inside a regular Docker registry, which will be handled outside of the system. 
+
+Conda dependencies are stored in Conda channel (where referenced packages are stored), which will be handled/setuped separately. (Popular conda channels are `default` and `conda-forge`)
+
+For more detailed discussion about storage-related implementations, please refer to [storage-implementation](./storage-implementation,md).
+
+## How to implement to make user can easily use Submarine environments? 
+
+We like simplicities, and we don't want to leak complexities of implementations to the users. To make it happen, we have to do some works to hide complexities. 
+
+There're two primary uses of environments: experiments and notebook, for both of them, users should not do works like explictily call `conda active $env_name` to active environments. To make it happen, what we can do is to include following parts in Dockerfile 
+
+```
+FROM ubuntu:18.04
+
+<Include whatever base-libraries like CUDA, etc.>
+
+<Make sure conda (with our preferred version) is installed>
+<Make sure Jupyter (with our preferred version) is installed>
+
+ENTRYPOINT ["/submarine-bootstrap.sh"]
```

Comment:
ENTRYPOINT ["/submarine-bootstrap.sh"] # This line is a sample code, the actual situation may change a little

Reply:
Updated doc and made it more clear.